### PR TITLE
Make terminal-host telemetry opt-in

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

`WithTerminal()` currently exports telemetry from its hidden terminal-host helpers whenever an OTLP endpoint is available. Their logs register `Aspire.TerminalHost` in the dashboard's telemetry resource list, including the Traces filter, even when no spans are emitted.

Terminal-host logs, metrics, and traces are now disabled by default. The existing `ShowTerminalHost = true` option both reveals the helper resources and enables their diagnostic telemetry when an OTLP endpoint is available. Parent-resource telemetry and terminal interaction are unchanged.

- Explicitly set a per-helper telemetry flag so inherited OTLP settings or diagnostic opt-in cannot enable hidden helpers.
- Only configure OTLP exporters and telemetry startup diagnostics for opted-in helpers.
- Preserve DCP's service name and instance ID so each helper's telemetry is associated with the correct resource and replica.
- Update option documentation and add focused coverage for telemetry gating, provider registration, and resource identity.

### User-facing usage

Use `.WithTerminal()` normally without exporting helper telemetry. To diagnose a terminal host, opt in with:

```csharp
resource.WithTerminal(options => options.ShowTerminalHost = true);
```

### Validation

- Hosting terminal configuration and failure-diagnostic tests: 53 passed, 1 platform-specific skip.
- TerminalHost telemetry and lifecycle tests: 20 passed in the selected run.
- Ran a CLI-launched AppHost against locally built binaries with a default terminal and a two-replica project using `ShowTerminalHost = true`. All producers were alive, terminal input round-tripped, and only the two opted-in helpers exported logs. Exported service names and instance IDs matched DCP's resource identities despite ambient telemetry settings.
- The daily-build reproduction showed helper logs and a telemetry resource entry, not actual trace rows; both the baseline and fixed runs had zero traces and spans.

The existing `DownstreamPrimaryResizeIsForwardedUpstreamAsRawResizeFrame` test timed out in an isolated run waiting for a resize frame and was excluded from the successful TerminalHost selection. It calls the unchanged instance `RunAsync` path rather than the modified telemetry entry point; it was not changed or quarantined.

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No

### CODE REVIEW

**Comment by mitchdenny** [comment]
# PR Testing Report

## PR Information

- **PR:** #20005 - Make terminal-host telemetry opt-in
- **Head tested:** `78972bbca0b7b3bbe9500ba097f57d5de605048d`
- **Tested:** 2026-09-09, macOS ARM64
- **Result:** The targeted telemetry behavior is verified against the PR artifacts. Restart recovery works with a new terminal attachment; an existing attachment did not survive restart.

## Artifact Version Verification

- **Installed CLI:** `13.6.0-pr.20005.g78972bbc`, matching the PR head.
- **Hosting package:** `Aspire.Hosting/13.6.0-pr.20005.g78972bbc`, confirmed in the generated AppHost's restored assets.
- **Terminal-host executable:** The `managed/aspire-managed` executable inside that PR CLI's versioned bundle, confirmed from running resource metadata. No local-source binary override was used.
- **Artifacts:** [CI run 34320297397](https://github.com/microsoft/aspire/actions/runs/34320297397).
- **Status:** Verified. The first installation preceded artifact publication; the unchanged dogfood installer succeeded after the macOS archive build completed.

## Changes Analyzed

| File | Changed behavior |
|---|---|
| `src/Aspire.Hosting/TerminalResourceBuilderExtensions.cs` | Per-helper opt-in flag and conditional OTLP configuration |
| `src/Aspire.Hosting/ApplicationModel/TerminalAnnotation.cs` | `ShowTerminalHost` documentation |
| `src/Aspire.TerminalHost/TerminalHostApp.cs` | Telemetry initialization gate and service identity preservation |
| `src/Aspire.TerminalHost/TerminalHostTelemetry.cs` | Telemetry documentation |
| `src/Shared/KnownConfigNames.cs` | Shared internal opt-in setting |
| `tests/Aspire.Hosting.Tests/WithTerminalTests.cs` | Hosting environment coverage |
| `tests/Aspire.TerminalHost.Tests/TerminalHostTelemetryTests.cs` | Provider gating and identity coverage |

Categories: Hosting, shared terminal-host runtime, documentation, and tests. No dashboard implementation, templates, client integrations, VS Code extension, or CI infrastructure changes.

## Test Setup

Used an isolated local install and three fresh projects generated with the PR template hive, not the earlier source-build fixtures.

```bash
curl -fsSL https://raw.githubusercontent.com/microsoft/aspire/main/eng/scripts/get-aspire-cli-pr.sh \
  | bash -s -- 20005 --install-path "$testDir/install" --skip-path --skip-extension
```

Template creation explicitly selected the PR hive and version, C#, `--localhost-tld false`, `--suppress-agent-init`, and `--non-interactive`. The starter also used `--test-framework None --use-redis-cache false`. AppHosts ran with `--isolated`; follow-up commands specified `--apphost`.

The experimental `aspire terminal` commands were enabled only in the temporary projects:

```bash
"$cli" config set features.terminalCommandsEnabled true --non-interactive
```

Terminal input was exercised through Hex1b; dashboard assertions and screenshots used headless Edge. No global CLI settings or repository files were changed.

## Test Scenarios Executed

### 1. Default helpers, inherited settings, and parent telemetry

**Coverage:** Happy path and negative configuration case. **Status:** Passed.

Added default `WithTerminal()` to the starter API service and a `/bin/cat` executable. Launched with an ambient OTLP endpoint, service name, instance ID, `ASPIRE_TERMINAL_HOST_TELEMETRY_ENABLED=true`, and debug logging.

- Both helpers were running but explicitly received telemetry opt-in `false`; no helper OTLP endpoint was injected.
- The visible resource list was exactly `apiservice`, `default-terminal`, and `webfrontend`.
- The actual dashboard Traces filter contained exactly `(All)`, `apiservice`, and `webfrontend`. No helper or `Aspire.TerminalHost` resource appeared.
- Structured logs came from the application services. Two successful `GET /weatherforecast` traces remained visible, with parent spans also retrieved by trace ID.
- Parent runtime metrics remained available, including positive `dotnet.process.memory.working_set` samples.
- Input round-tripped through the default terminal.

**Evidence:** `default-traces-filter.png`, `default-terminal.svg`, `parent-metrics.png`, `default-traces-assertions.txt`, `parent-metric-values.txt`, `runtime-evidence.json`.

### 2. Opted-in diagnostic replicas

**Coverage:** Happy path and replica identity. **Status:** Passed.

Created a real console project with two replicas and:

```csharp
.WithReplicas(2)
.WithTerminal(options => options.ShowTerminalHost = true)
```

Both producers were alive, both helpers appeared in the resource list, and the workload responded to terminal input with `echo: diagnostic-artifact-probe`.

Exported log resource attributes exactly matched DCP's injected identities:

| Service name | Service instance ID |
|---|---|
| `diagnostic-terminal-terminalhost-0` | `rfvxqgbz` |
| `diagnostic-terminal-terminalhost-1` | `ejtfkgxn` |

The dashboard Traces and Metrics selectors contained the two helper names. Live terminal-host metric instruments were present, and the accessible metric table showed `aspire.terminalhost.consumer.connections = 2` for replica 1.

**Evidence:** `diagnostic-resources.png`, `diagnostic-terminal.svg`, `diagnostic-traces-filter.png`, `diagnostic-metrics.png`, `helper-metric-values.txt`, `runtime-evidence.json`.

### 3. Replica restart and terminal recovery

**Coverage:** Lifecycle/recovery. **Status:** Restart and reattachment passed; continuous attachment did not.

Restarted both concrete project instances using `aspire resource <instance> restart`. Both became running again, their process IDs changed, and terminal status reported one restart for each.

An exploratory assertion that the existing CLI attachment would continue accepting input failed: the attachment disconnected during restart. A fresh attachment successfully processed `echo: after-restart-probe`.

This run establishes recovery by reattachment, not transparent reconnection. It did not compare attachment continuity against the base branch, so this observation is not classified as a regression introduced by this PR.

**Evidence:** `restarted-terminal.svg`, before/after process IDs and attachment outcomes in `runtime-evidence.json`.

### 4. No dashboard and missing workload

**Coverage:** Boundary and expected failure. **Status:** Passed.

A fresh AppHost disabled the dashboard and cleared its OTLP endpoint configuration. A default terminal still started and round-tripped input.

A second resource referenced a deliberately nonexistent executable. It visibly entered `FailedToStart`; `aspire wait ... --status up` returned exit code **18**, rather than silently succeeding. Its helper remained waiting for the nonexistent producer.

**Evidence:** `no-endpoint-terminal.svg`, `failed-workload-wait.txt`, `runtime-evidence.json`.

### 5. Standalone helper opted in without an endpoint

**Coverage:** Telemetry initialization boundary. **Status:** Passed.

Launched the actual packaged `aspire-managed terminalhost` with telemetry opt-in `true`, trace-level logging, and no `OTEL_EXPORTER_OTLP_ENDPOINT`. All three Unix sockets bound, the control socket accepted a connection, and telemetry startup stderr remained empty. The process was then stopped.

**Evidence:** `runtime-evidence.json`.

## Coverage Limits

The helpers produced logs and metrics, not lifecycle activities. Therefore this run verifies their Traces resource-list behavior, but does not claim to have observed exported helper spans. Tracer-provider gating and identity are covered by the focused source tests added in this PR; those earlier local unit results are not represented as new artifact-test runs here.

## CI Observations

The PR's TerminalHost tests passed on [Linux](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102365740437) and [Windows](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102365739750).

Two other jobs initially failed and subsequently passed on retry:

- [Deno debugger E2E initial failure](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102368001335): timed out after 240 seconds waiting for a TypeScript breakpoint. [Retry](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102372753777).
- [Radius Windows initial failure](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102365740663): `WaitForSealedSecretSynced_StatusMatchesButSecretAbsent_KeepsWaitingThenTimesOut` failed `Assert.True` at `SealedSecretApplyStepTests.cs:513`. [Retry](https://github.com/microsoft/aspire/actions/runs/34320297397/job/102372733061).

At 07:19:41 UTC the final check snapshot was **309 passed, 7 skipped, none pending or failing**. No code changes, quarantines, issue creation, or manual CI reruns were performed during this testing task.

## Summary

| Scenario | Result |
|---|---|
| Default helper suppression, including inherited opt-in | Passed |
| Parent logs, HTTP traces, and runtime metrics | Passed |
| Diagnostic replica identity and live helper metrics | Passed |
| Terminal input before restart and after reattachment | Passed |
| Existing CLI attachment surviving restart | Did not survive; recorded observation |
| Dashboard-free terminal and visible workload failure | Passed |
| Standalone opted-in helper without OTLP endpoint | Passed |

## Overall Result

**Targeted telemetry behavior verified against the matching PR artifacts.** The restart attachment observation prevents claiming uninterrupted terminal continuity. Earlier broader CI failures passed on retry, and the final CI snapshot had no pending or failing checks.

## Artifacts and Cleanup

Sanitized evidence is retained with the local testing report: five dashboard PNG screenshots, four terminal SVG screenshots, runtime JSON, assertion output, fixture source, installer logs, and CI logs. Dashboard login URLs, API keys, and raw environment exports are not included in this report.

All test AppHosts, terminal viewers, the standalone helper, and headless browser sessions were stopped. The temporary install/project workspace was removed. The repository remains unchanged and the PR remains a draft.

File: src/Aspire.TerminalHost/TerminalHostApp.cs [hunk 1]
```diff
@@ -415,6 +352,39 @@ await Console.Error.WriteLineAsync(
         }
     }
 
+    internal static HostApplicationBuilder? CreateTelemetryHostBuilder(IConfiguration configuration)
+    {
+        // An inherited OTLP endpoint must not turn a hidden implementation detail into
+        // a telemetry resource. The AppHost explicitly sets this flag from ShowTerminalHost.
+        if (!configuration.GetValue<bool>(KnownConfigNames.TerminalHostTelemetryEnabled) ||
```
Comment (JamesNK) [inline src/Aspire.TerminalHost/TerminalHostApp.cs]: Could this fail closed for malformed values? `GetValue<bool>` throws before the app's normal error handling, so a typo in standalone diagnostic configuration terminates the terminal host with an unhandled `InvalidOperationException` (I reproduced this with `not-a-bool`). Since telemetry is opt-in, using `bool.TryParse` or the shared `GetBool` behavior and treating invalid values as disabled would keep the terminal available; an invalid-value test would cover the boundary.

**Review by JamesNK — APPROVED** [review]
Approved. I left one non-blocking robustness comment about failing closed when the telemetry opt-in environment variable is malformed. Focused tests and direct OTLP proof otherwise confirmed the intended hidden/visible behavior and per-replica resource identity.

File: src/Aspire.TerminalHost/TerminalHostApp.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.TerminalHost/TerminalHostApp.cs]: Fixed in 6ac0f05050. The opt-in now uses `bool.TryParse`, so malformed values leave telemetry disabled instead of throwing during terminal-host startup. Added cases for an empty string, whitespace, `not-a-bool`, and `1`; all four reproduced the exception before the change, and all 12 telemetry tests now pass.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Aspire.TerminalHost/TerminalHostApp.cs [hunk 2]
```diff
@@ -415,6 +352,41 @@ await Console.Error.WriteLineAsync(
         }
     }
 
+    internal static HostApplicationBuilder? CreateTelemetryHostBuilder(IConfiguration configuration)
+    {
+        // An inherited OTLP endpoint must not turn a hidden implementation detail into
+        // a telemetry resource. The AppHost explicitly sets this flag from ShowTerminalHost.
+        // Malformed values (e.g. "not-a-bool") must leave telemetry disabled, not stop the terminal.
+        if (!bool.TryParse(configuration[KnownConfigNames.TerminalHostTelemetryEnabled], out var telemetryEnabled) ||
```
Comment (JamesNK) [inline src/Aspire.TerminalHost/TerminalHostApp.cs]: nit: I think you should use https://github.com/microsoft/aspire/blob/cc5152608322e2eb10374aafafea38caf001f7d4/src/Shared/IConfigurationExtensions.cs#L42

Just reference the shared file from the terminal host project.

Comment (mitchdenny) [reply src/Aspire.TerminalHost/TerminalHostApp.cs]: Done in ad937d4940. TerminalHost now links the shared `IConfigurationExtensions.cs` and calls `GetBool(..., defaultValue: false)`. I extended the existing CLI/dashboard attribute shim to keep TerminalHost independent of the Hosting assembly. The regression cases still cover empty and malformed input, and now cover the shared helper's numeric semantics (`0` disables; `1` and `-1` enable). All 14 telemetry tests pass.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).
